polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -248,30 +256,33 @@ mlir::Attribute MLIRScanner::InitializeValueByInitListExpr(mlir::Value ToInit,
 
         mlir::Value Next;
         if (auto MT = ToInit.getType().dyn_cast<MemRefType>()) {
-          auto Shape = std::vector<int64_t>(MT.getShape());
+          llvm::SmallVector<int64_t> Shape(MT.getShape());
           assert(!Shape.empty());
           Shape[0] = ShapedType::kDynamic;
 
-          if (MT.getElementType()
-                  .isa<mlir::sycl::AccessorType,
-                       mlir::sycl::AccessorImplDeviceType,
-                       mlir::sycl::ArrayType, mlir::sycl::ItemType,
-                       mlir::sycl::NdItemType, mlir::sycl::GroupType,
-                       mlir::sycl::VecType>()) {
-            llvm_unreachable("not implemented yet");
-          }
-
           mlir::Type ET;
           if (auto ST =
                   MT.getElementType().dyn_cast<mlir::LLVM::LLVMStructType>()) {
             ET = mlir::MemRefType::get(Shape, ST.getBody()[I],
                                        MemRefLayoutAttrInterface(),
                                        MT.getMemorySpace());
-          } else if (auto ST = MT.getElementType()
-                                   .dyn_cast<mlir::sycl::ItemBaseType>()) {
-            ET = mlir::MemRefType::get(Shape, ST.getBody()[I],
-                                       MemRefLayoutAttrInterface(),
-                                       MT.getMemorySpace());
+          } else if (sycl::isSYCLType(MT.getElementType())) {
+            mlir::Type ElemTy = MT.getElementType();
+            ET =
+                TypeSwitch<mlir::Type, mlir::MemRefType>(ElemTy)
+                    .Case<mlir::sycl::IDType, mlir::sycl::RangeType>(
+                        [&](auto T) {
+                          return mlir::MemRefType::get(
+                              Shape, MT.getElementType(),
+                              MemRefLayoutAttrInterface(), MT.getMemorySpace());
+                        })
+                    .Case<mlir::sycl::ItemBaseType>([&](auto ElemTy) {
+                      return SYCLGetElementType<decltype(ElemTy)>(
+                          ElemTy, I, Shape, MT.getMemorySpace());
+                    })
+                    .Default([](mlir::Type T) -> mlir::MemRefType {
```

Comment:
(mlir::Type T) -> auto

Reply:
have you tried to return nullptr and remove the `-> mlir::MemRefType` ?

Reply:
I'm wondering whether we should also print the type before llvm_unreachable ?

Reply:
> have you tried to return nullptr and remove the -> mlir::MemRefType ?

It works, but why do you prefer that?

Reply:
because the Default case should return something (like the other cases) and making it return nullptr does that.

Reply:
[nit]: use `Ty` not `T` because the case at line 270 does. Also use auto to be consistent with the same line.

Reply:
I prefer to keep `-> mlir::MemRefType` without return, as we actually don't want to return, and we only want to add the return to avoid the compile error, which `-> mlir::MemRefType` does the same.